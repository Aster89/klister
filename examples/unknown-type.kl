#lang "n-ary-app.kl"

(import (shift "n-ary-app.kl" 1))
(import (shift "quasiquote.kl" 1))
(import "pair-datatype.kl")

(define-macros
  ([gotta-be-maybe
    (lambda (stx)
      (syntax-case stx
        [(list (_ e))
         (pure (quasiquote/loc stx (with-unknown-type (A) (the (Maybe A) ,e))))]))]
   [gotta-be-maybes
    (lambda (stx)
      (syntax-case stx
        [(list (_ e1 e2))
         (pure
          (quasiquote/loc stx
            (with-unknown-type (A)
              (pair (the (Maybe A) ,e1)
                    (the (Maybe A) ,e2)))))]))]))

(example (gotta-be-maybe (nothing)))
(example (gotta-be-maybe (just 'here)))
(example (gotta-be-maybe (just (lambda (x) x))))


(example (gotta-be-maybes (just (lambda (x) x)) (nothing)))
