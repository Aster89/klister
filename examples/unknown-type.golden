(nothing) : ∀(α : *). (Maybe α)
(just #[unknown-type.kl:24.33-24.37]<here>) : (Maybe Syntax)
(just #<closure>) : ∀(α : *). (Maybe (α → α))
(pair (just #<closure>) (nothing)) : ∀(α : *). (Pair (Maybe (α → α)) (Maybe (α → α)))
